caikit/core/module_backends/base.py

```diff
@@ -54,3 +59,67 @@ def start(self):
     def stop(self):
         """Function to stop a distributed backend. This function
         should set self._started variable to False"""
+
+
+class SharedTrainBackendBase(BackendBase, abc.ABC):
+    """Interface for a backend that can perform train on any given module
+
+    A Shared backend is one that treats the given module as a black box and
+    delegates the execution of that module's functionality to an alternate
+    execution engine.
+    """
+
+    @abc.abstractmethod
+    def train(self, module_class: Type[ModuleBase], *args, **kwargs) -> ModuleBase:
```

Comment:
Just thinking out the train use-case:
1. Would train data massaging / packaging also be responsibility of train backend impl class? Like in case of training in a torch cluster, if it requires data in torch data loaders, would the conversion from potentially stream in `caikit.runtime` to torch data loader be responsibility of impl of `SharedTrainBackend`?
2. Would we always be running `caikit.module` based training code in the backend? Or are we going to support non-caikit module training in backend as well?
3. Does error / exception handling responsibility of train function? Considering below `TODO` around async training, since `train` here can return future, would we need to expose another semantics for checking on the status, fetching error etc?
4.  How would the `.save` be handled and the case of automatic loading of the model, like the `auto_load_trained_model` case?

Reply:
>Would train data massaging / packaging also be responsibility of train backend impl class? Like in case of training in a torch cluster, if it requires data in torch data loaders, would the conversion from potentially stream in caikit.runtime to torch data loader be responsibility of impl of SharedTrainBackend?

I have a hard time imagining a `SharedTrainBackend` that wraps running in a `torch` cluster since that would only work for module implementations that are backed by `torch`. I would expect such usages to be non-shared backends that require backend-specific implementations for a remote `torch` cluster. Realistically, the data conversions would probably be very similar across models, so the backend could expose any utility functions needed to make this code reusable.

In general, though, for a truly `Shared` train backend, yes, any data massaging would be the responsibility of the backend itself. For example, a shared train backend that delegates to a remote instance of `caikit.runtime` would need to serialize the python args into a train request `grpc` call.

>Would we always be running caikit.module based training code in the backend? Or are we going to support non-caikit module training in backend as well?

Good question. I think the baseline is that we would always have any code encapsulated as a `caikit` module. It could be totally fair game to have a generic module that allows for arbitrary training against a given platform, though. Again, something like this would probably not be a `Shared` backend, though, since the only generic interface that a `SharedTrainBackend` can rely on is `ModuleBase`.

>Does error / exception handling responsibility of train function? Considering below TODO around async training, since train here can return future, would we need to expose another semantics for checking on the status, fetching error etc?

Currently this does not return a future, but we absolutely will need to have careful error handling semantics down the line. Since the goal of a `SharedTrainBackend` is to encapsulate platform-specific concerns for a given backend platform, I think we will likely want to define a set of standard training error flows and make sure that all `SharedTrainBackend` implementations follow them.

>How would the .save be handled and the case of automatic loading of the model, like the auto_load_trained_model case?

This will be the crux of the `orchestration` design once we get there. We'll need to make sure that the train and load backends are able to interoperate cleanly to enable `auto_load_trained_model` semantics, even when the model is being trained and loaded on divergent backends. Certainly some design work to do here to figure out how to validate configurations and perform the model handoff from training to inference.

>From runtime perspective, can a shared backend in itself do distribution of training (internal distribution) and manage its own cluster?

By definition, a `SharedTrainBackend` is performing `external` distribution (the module is the atomic unit being distributed). Hypothetically a given shared backend _could_ also be used for `internal` distribution (distribution of parts of the workload inside a given module). For example, I could imagine a generic `ray` backend that holds the connection to a cluster and can perform distribution of modules as `remote` (`actor`, `job`?) primitives as a `SharedTrainBackend`, but also allows individual `ray`-based module implementations to grab the client and perform internal distribution.

As for "and manage its own cluster," this is an interesting question. My gut feeling is that or anything that will be running in a truly "remote" sense, the deployment topology should live outside the backend implementation to avoid mixing runtime and deploy-time concerns. That said, you could certainly imagine something like how [`caikit-tgis-backend` runs its subprocess](https://github.com/caikit/caikit-tgis-backend/blob/main/caikit_tgis_backend/tgis_backend.py#L82) for cluster-based backends that want to be able to operate in "build your own" mode.

Reply:
Thanks Gabe.  

I think we would need to document above definitions as "parameters" for designing shared backends, so that not everything ends up being one-off shared backends and the difference between shared and non-shared backend is evident.

Reply:
100% agree. We probably need a pretty robust ADR around this. I can start writing one up today, but we need to sort out numbering at some point @joerunde